### PR TITLE
Add a function to Get the fully-assembled operator in a single SparseMatrix

### PR DESCRIPTION
**Pull request by YohannDudouit**
This PR adds a function to get a single `SparseMatrix` corresponding to the fully-assembled operator:
```c++
BilinearForm b;
...
b.SetAssemblyLevel(AssemblyLevel::FULL);
SparseMatrix &A = b.SpMat();
```
ping: @artv3 

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1732](https://github.com/mfem/mfem/pull/1732) | @YohannDudouit | @v-dobrev | @artv3 + @acfisher | 12/09/20 | 01/19/21 | ⌛due 01/26/21 | |

### CODE REVIEW

File: examples/fa-test.cpp [hunk 1]
```diff
@@ -0,0 +1,806 @@
...
+   //Reference code -- legacy assembly 
+   ParBilinearForm *k_ref = new ParBilinearForm(fes);
+   k_ref->AddDomainIntegrator(new ConvectionIntegrator(velocity, -1.0));
+   k_ref->AddInteriorFaceIntegrator(
+      new TransposeIntegrator(new DGTraceIntegrator(velocity, 1.0, -0.5)));
+   
+   k_ref->Finalize(skip_zeros);
+   k_ref->KeepNbrBlock();
+
+   double iError(0), jError(0), dataError(0);
```
Comment (artv3) [inline examples/fa-test.cpp:378]: @YohannDudouit , one way to check if the routine works is to compare the CSR data structure to the legacy version of the code (no boundary conditions imposed here). Below I check the I, J, Data arrays. Currently this example  segfaults, but getting this to work would be a good starting point. Ideally, we could then reuse the sparse matrix to repopulate on the GPU - for time dependent coefficient.

Comment (YohannDudouit) [reply examples/fa-test.cpp:378]: @artv3 The test you wrote is SegFaulting in something that has nothing to do with `GetFullAssemblySparseMatrix`, can you fix it?

Comment (artv3) [reply examples/fa-test.cpp]: Yes I can! Fixed and Pushed :-). 
I added some print statements to look at the sizes of the array. I see some differences between J, and Data: 
 lrun -n 1 ./fa-test gives 
```
size of I k_ref 3073 A 3073 
size of J k_ref 245760 A 61440 
size of Data k_ref 245760 A 61440 
```

Comment (YohannDudouit) [reply examples/fa-test.cpp]: I think your test doesn't really work, in this example the `SparseMatrix` are the same but your error is high:
```
 ~/Projects/mfem/examples/> ./fa-test -fa -m ../data/inline-quad.mesh -o 1
Options used:
   --mesh ../data/inline-quad.mesh
   --problem 0
   --refine-serial 0
   --refine-parallel 0
   --order 1
   --no-partial-assembly
   --no-element-assembly
   --full-assembly
   --device cpu
   --ode-solver 4
   --t-final 10
   --time-step 0.01
   --visualization
   --no-visit-datafiles
   --no-paraview-datafiles
   --no-adios2-streams
   --ascii-datafiles
   --visualization-steps 5
Device configuration: cpu
Memory configuration: host-std
Number of unknowns: 4
A=
[row 0]
 (0,0.23230781) (1,-0.087970241) (2,-0.028183663) (3,-0.1161539)
[row 1]
 (0,0.18419529) (1,-0.039857719) (2,0.019928859) (3,-0.16426643)
[row 2]
 (0,0.16426643) (1,-0.019928859) (2,0.039857719) (3,-0.18419529)
[row 3]
 (0,0.1161539) (1,0.028183663) (2,0.087970241) (3,-0.23230781)

A_ref=
[row 0]
 (3,-0.1161539) (2,-0.028183663) (1,-0.087970241) (0,0.23230781)
[row 1]
 (3,-0.16426643) (2,0.019928859) (1,-0.039857719) (0,0.18419529)
[row 2]
 (3,-0.18419529) (2,0.039857719) (1,-0.019928859) (0,0.16426643)
[row 3]
 (3,-0.23230781) (2,0.087970241) (1,0.028183663) (0,0.1161539)

size of I k_ref 5 A 5 
size of J k_ref 16 A 16 
size of Data k_ref 16 A 16 
iError 0 
jError 8.94427 
dataError 1 
error too high 
```

Comment (YohannDudouit) [reply examples/fa-test.cpp]: Looking at a 2 elements mesh, the matrices are still the same, but the legacy sparse matrix seems to be storing a lot of zeros:
```
 ~/Projects/mfem/examples/> ./fa-test -fa -m ../data/inline-quad.mesh -o 1
Options used:
   --mesh ../data/inline-quad.mesh
   --problem 0
   --refine-serial 0
   --refine-parallel 0
   --order 1
   --no-partial-assembly
   --no-element-assembly
   --full-assembly
   --device cpu
   --ode-solver 4
   --t-final 10
   --time-step 0.01
   --visualization
   --no-visit-datafiles
   --no-paraview-datafiles
   --no-adios2-streams
   --ascii-datafiles
   --visualization-steps 5
Device configuration: cpu
Memory configuration: host-std
Number of unknowns: 8
A=
[row 0]
 (0,0.18419529) (1,-0.1120265) (2,0.019928859) (3,-0.092097643)
[row 1]
 (0,0.16013902) (1,-0.087970241) (2,0.043985121) (3,-0.1161539)
 (4,0) (6,0)
[row 2]
 (0,0.1161539) (1,-0.043985121) (2,0.087970241) (3,-0.16013902)
[row 3]
 (0,0.092097643) (1,-0.019928859) (2,0.1120265) (3,-0.18419529)
 (4,0) (6,0)
[row 4]
 (4,-0.087970241) (5,-0.1120265) (6,-0.1161539) (7,-0.092097643)
 (1,0.27216553) (3,0.13608276)
[row 5]
 (4,0.16013902) (5,-0.087970241) (6,0.043985121) (7,-0.1161539)
[row 6]
 (4,-0.019928859) (5,-0.043985121) (6,-0.18419529) (7,-0.16013902)
 (1,0.13608276) (3,0.27216553)
[row 7]
 (4,0.092097643) (5,-0.019928859) (6,0.1120265) (7,-0.18419529)

A_ref=
[row 0]
 (7,0) (6,0) (5,0) (4,0)
 (3,-0.092097643) (2,0.019928859) (1,-0.1120265) (0,0.18419529)
[row 1]
 (7,0) (6,0) (5,0) (4,0)
 (3,-0.1161539) (2,0.043985121) (1,-0.087970241) (0,0.16013902)
[row 2]
 (7,0) (6,0) (5,0) (4,0)
 (3,-0.16013902) (2,0.087970241) (1,-0.043985121) (0,0.1161539)
[row 3]
 (7,0) (6,0) (5,0) (4,0)
 (3,-0.18419529) (2,0.1120265) (1,-0.019928859) (0,0.092097643)
[row 4]
 (3,0.13608276) (2,0) (1,0.27216553) (0,0)
 (7,-0.092097643) (6,-0.1161539) (5,-0.1120265) (4,-0.087970241)
[row 5]
 (3,0) (2,0) (1,0) (0,0)
 (7,-0.1161539) (6,0.043985121) (5,-0.087970241) (4,0.16013902)
[row 6]
 (3,0.27216553) (2,0) (1,0.13608276) (0,0)
 (7,-0.16013902) (6,-0.18419529) (5,-0.043985121) (4,-0.019928859)
[row 7]
 (3,0) (2,0) (1,0) (0,0)
 (7,-0.18419529) (6,0.1120265) (5,-0.019928859) (4,0.092097643)

size of I k_ref 9 A 9 
size of J k_ref 64 A 40 
size of Data k_ref 64 A 40 
iError 42.332 
jError nan 
dataError 1.17465 
error too high 
```

Comment (YohannDudouit) [reply examples/fa-test.cpp]: I also added:
```
   k->AddBdrFaceIntegrator(
      new TransposeIntegrator(new DGTraceIntegrator(velocity, 1.0, -0.5)));
```
and the `SparseMatrix` are still the same.

Comment (YohannDudouit) [reply examples/fa-test.cpp]: (it's only if essential dofs, I think, that the results would be different)

Comment (artv3) [reply examples/fa-test.cpp]: Interesting (I forgot they might be storing a different number of zeros -- although I was hoping in CSR we wouldn't store zeros... ) ! I turned on skip_zeros (skip_zeros=1) in my last push, and I did see some differences in array sizes. But if they are still the same matrices but with different numbers of zeros in it, then that is fine. We can just do a convergence test with a periodic mesh and if it checks out we can confirm it is correct.

Comment (artv3) [reply examples/fa-test.cpp]: Also, I just noticed the column indices are sorted in the new implementation, so of course they won't agree unless we sort them...

Comment (artv3) [reply examples/fa-test.cpp]: Idea! Perhaps we can filter out zeros and sort the columns when comparing the old and new matrices.

Comment (YohannDudouit) [reply examples/fa-test.cpp]: I don't really sort them, I happen to insert the values in order, this might not hold on GPU, due to concurrent access.

Comment (YohannDudouit) [reply examples/fa-test.cpp]: I'm not sure this will be true for meshes with more elements.

Comment (YohannDudouit) [reply examples/fa-test.cpp]: It's actually already not holding:
```
[row 6]
 (4,-0.019928859) (5,-0.043985121) (6,-0.18419529) (7,-0.16013902)
 (1,0.13608276) (3,0.27216553)
```

Comment (YohannDudouit) [reply examples/fa-test.cpp]: You would need to sort both of them.

**Comment by YohannDudouit** [comment]
@artv3 I think it works in parallel now. I noticed a difference between what I return and what `SpMat()` returns, I have more rows, i.e. I have rows for interior + shared dofs, whereas `SpMat()` has only columns for interior + shared dofs, but only interior dofs rows.

**Comment by YohannDudouit** [comment]
@v-dobrev Do you have a preference in term of design for this `SparseMatrix GetFullAssemblySparseMatrix(BilinearForm &a)` function, we can change everything about it, the name, the file location, etc...

**Comment by artv3** [comment]
> @artv3 I think it works in parallel now. I noticed a difference between what I return and what `SpMat()` returns, I have more rows, i.e. I have rows for interior + shared dofs, whereas `SpMat()` has only columns for interior + shared dofs, and only interior dofs rows.

Thanks @YohannDudouit! Will try this out. I'm not sure I understand your comment, could it be the matrix is transposed?

**Comment by YohannDudouit** [comment]
> > @artv3 I think it works in parallel now. I noticed a difference between what I return and what `SpMat()` returns, I have more rows, i.e. I have rows for interior + shared dofs, whereas `SpMat()` has only columns for interior + shared dofs, and only interior dofs rows.
> 
> Thanks @YohannDudouit! Will try this out. I'm not sure I understand your comment, could it be the matrix is transposed?

@artv3 I just did what you asked me, you said you wanted the blocks associated with the shared dofs, this is what this is.

**Comment by artv3** [comment]
> > > @artv3 I think it works in parallel now. I noticed a difference between what I return and what `SpMat()` returns, I have more rows, i.e. I have rows for interior + shared dofs, whereas `SpMat()` has only columns for interior + shared dofs, and only interior dofs rows.
> > 
> > 
> > Thanks @YohannDudouit! Will try this out. I'm not sure I understand your comment, could it be the matrix is transposed?
> 
> @artv3 I just did what you asked me, you said you wanted the blocks associated with the shared dofs, this is what this is.

Ah I see, gotcha. thanks.

**Comment by YohannDudouit** [comment]
@artv3 I changed `GetFullAssemblySparseMatrix` to return a rectangular `SparseMatrix` according to what we discussed.

File: fem/bilinearform.hpp [hunk 2]
```diff
@@ -292,15 +292,22 @@ class BilinearForm : public Matrix
...
-      MFEM_VERIFY(mat, "mat is NULL and can't be dereferenced");
-      return *mat;
+      if (GetAssemblyLevel() == AssemblyLevel::LEGACYFULL)
+      {
+         MFEM_VERIFY(mat, "mat is NULL and can't be dereferenced");
+         return *mat;
+      }
+      else
+      {
+         mat = mat ? mat : GetFullAssemblySparseMatrix(*this);
```
Comment (artv3) [inline fem/bilinearform.hpp:308]: nice!

**Comment by tzanio** [comment]
There is an ongoing discussion on this topic, so the PR is not ready for a formal review yet... (@YohannDudouit, correct me if I'm wrong)

**Comment by artv3** [comment]
make style :)

**Comment by artv3** [comment]
I hit a couple of errors when building with MPI + CUDA: 

```
fem/bilinearform_ext.cpp(955): error: The enclosing parent function ("DGMult") for an extended __device__ lambda cannot have private or protected access within its class

fem/bilinearform_ext.cpp(961): error: The enclosing parent function ("DGMult") for an extended __device__ lambda cannot have private or protected access within its class

fem/bilinearform_ext.cpp(972): error: The enclosing parent function ("DGMult") for an extended __device__ lambda cannot have private or protected access within its class

fem/bilinearform_ext.cpp(1016): error: The enclosing parent function ("DGMultTranspose") for an extended __device__ lambda cannot have private or protected access within its class

fem/bilinearform_ext.cpp(1022): error: The enclosing parent function ("DGMultTranspose") for an extended __device__ lambda cannot have private or protected access within its class

fem/bilinearform_ext.cpp(1033): error: The enclosing parent function ("DGMultTranspose") for an extended __device__ lambda cannot have private or protected access within its class

```

**Comment by YohannDudouit** [comment]
@artv3 Sorry, I forgot that we can't use private methods...

File: fem/bilinearform.hpp [hunk 3]
```diff
@@ -292,8 +294,7 @@ class BilinearForm : public Matrix
    /// Returns a const reference to the sparse matrix.
    const SparseMatrix &SpMat() const
    {
-      MFEM_VERIFY(mat, "mat is NULL and can't be dereferenced");
-      return *mat;
+      return const_cast<BilinearForm*>(this)->SpMat();
```
Comment (artv3) [inline fem/bilinearform.hpp:297]: Would it also be necessary to add the `` return const_cast<BilinearForm*>(this)->SpMat();`` to the non-const method below? Just for consistency?

Comment (YohannDudouit) [reply fem/bilinearform.hpp:297]: What this code does is that it uses the non-const version to implement the const version, this is the standard way to implement const and non-const versions of a method (to avoid code duplication). But I can revert this, it's not important any more because I don't change any more the non-const version.

File: tests/unit/fem/test_sparse_matrix.cpp [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+   Mesh mesh(input, 1, 1);
+   mesh.EnsureNodes();
+   mesh.UniformRefinement();
+   ParMesh pmesh(MPI_COMM_WORLD, mesh);
+   int dim = mesh.Dimension();
+
+   FiniteElementCollection *fec;
+   if (pb == Problem::Convection)
+   {
+      fec = new L2_FECollection(order, dim, BasisType::GaussLobatto);
```
Comment (artv3) [inline tests/unit/fem/test_sparse_matrix.cpp:99]: Could we also add a test for ``BasisType::Positive`` ?

File: fem/bilinearform.hpp (hunk 3 above)
Comment (artv3) [reply fem/bilinearform.hpp:297]: Ah okay thanks. I think it is fine if it works for the non-const method as well.

File: tests/unit/fem/test_sparse_matrix.cpp (hunk 4 above)
Comment (YohannDudouit) [reply tests/unit/fem/test_sparse_matrix.cpp]: done.

**Comment by artv3** [comment]
Is this PR ready?

**Comment by YohannDudouit** [comment]
@artv3 This PR is ready for review. Supporting Bernstein basis had some consequences on mesh and pmesh.

**Review by artv3 — APPROVED** [review]
Solid work! I've been testing it in Remhos, I do see some run to run variations when using CUDA (still testing though not sure where it is coming from), no differences on the CPU. But I suspect it is not because of anything introduced here, I think we can look into it a bit more separately.

**Comment by artv3** [comment]
Just to add a note here, this PR also fixes a bug with PA and Bernstein basis.

**Comment by tzanio** [comment]
@acfisher -- a reminder that the review for this PR is due this week.

File: fem/fe_coll.cpp [hunk 5]
```diff
@@ -1991,7 +1991,7 @@ L2_FECollection::L2_FECollection(const int p, const int dim, const int btype,
       // Trace element use the default Gauss-Legendre nodal points for positive basis
       if (b_type == BasisType::Positive)
       {
-         Tr_Elements[Geometry::SEGMENT] = new L2_SegmentElement(p);
+         Tr_Elements[Geometry::SEGMENT] = new L2Pos_SegmentElement(p);
```
Comment (acfisher) [inline fem/fe_coll.cpp:1994]: Oh wow, these look like they were bugs.

File: tests/unit/fem/test_sparse_matrix.cpp [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+   x.Randomize(1);
+
+   k_test.SpMat().Mult(x,y_test);
+   k_ref.SpMat().Mult(x,y_ref);
+
+   y_test -= y_ref;
+
+   Vector result(y_test.HostReadWrite(), sizeEnd);
+
+   REQUIRE(result.Norml2() < 1.e-12);
```
Comment (acfisher) [inline tests/unit/fem/test_sparse_matrix.cpp:183]: This is a pretty loose tolerance for a single matvec.  Does the error get anywhere close to that?

**Review by acfisher — APPROVED** [review]

**Comment by v-dobrev** [comment]
Merged in `next` for testing...

**Comment by v-dobrev** [comment]
LGTM

**Comment by artv3** [comment]
Hi all just wanted to follow up on this PR. Is this next up to be merged into master?
